ray-on-gke/user/kuberay-tpu-webhook/main.go

```diff
@@ -0,0 +1,287 @@
+				return true
+			}
+		}
+		if r := container.Resources.Requests; r != nil {
+			if resource := r[tpuResourceName]; !resource.IsZero() {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+// check if request is for TPU multi-host
+func isTPUMultiHost(topology string) bool {
+	topologyVals :=  strings.Split(topology, "x")
+	values := make([]int, len(topologyVals))
+	for i := 0; i < len(values); i++ {
+		dim, err := strconv.Atoi(topologyVals[i])
+		if err != nil {
+			klog.Fatalf("Invalid topology value: %s", err)
+		}
+		values[i] = dim
+	}
+	vms := 1 // number VMs = number chips / 4
+	if len(values) == 3 {
+		// TPU v4
+		vms = max(values[0]*values[1]*values[2]/4, 1)
+	} else if len(values) == 2 {
+		// TPU v5e
+		// slices with 8 chips can be single or multi host
+		chips := values[0]*values[1]
+		vms = max(chips/4, 1)
+	}
+	return (vms > 1)
+}
+
+// unmarshal raycluster from admission request
+func extractRayCluster(admissionReview *admissionv1.AdmissionReview) (*ray.RayCluster, error) {
+	if admissionReview.Request.Kind.Kind != "RayCluster" {
+		return nil, fmt.Errorf("Expected RayCluster but got %s", admissionReview.Request.Kind.Kind)
+	}
+
+	rayCluster := ray.RayCluster{}
+	if err := json.Unmarshal(admissionReview.Request.Object.Raw, &rayCluster); err != nil {
+		return nil, err
+	}
+
+	return &rayCluster, nil
+}
+
+func genDNSHostnames(workerGroupSpec ray.WorkerGroupSpec) string {
+	numWorkers := int(*workerGroupSpec.Replicas)
+	workerGroupName := workerGroupSpec.GroupName
+	serviceName := "tpu-worker-group-svc" // TODO: get headless service from workergroup spec
+	hostNames := make([]string, numWorkers)
+	for j := 0; j < numWorkers; j++ {
+		hostNames[j] = fmt.Sprintf("%s-%d.%s", workerGroupName, j, serviceName)
+	}
+	return strings.Join(hostNames, ",")
+}
+
+func injectHostnames(hostNames string, workerGroupSpec ray.WorkerGroupSpec, workerGroupIndex int, patches *[]patch) {
+	for j := 0; j < len(workerGroupSpec.Template.Spec.Containers); j++ {
+		container := workerGroupSpec.Template.Spec.Containers[j]
+		if containerRequestingTPUs(container) {
+			hostNamesPatch := patch {
+				"op": "add",
+			}
+			path := fmt.Sprintf("/spec/workerGroupSpecs/%d/template/spec/containers/%d/env", workerGroupIndex, j)
+			tpuWorkerHostNames := corev1.EnvVar{
+				Name:  "TPU_WORKER_HOSTNAMES",
+				Value: hostNames,
+			}
+			if len(container.Env) == 0 {
+				hostNamesPatch["path"] = path
+				hostNamesPatch["value"] = []corev1.EnvVar{tpuWorkerHostNames}
+			} else {
+				hostNamesPatch["path"] = fmt.Sprintf("%s/-", path)
+				hostNamesPatch["value"] = tpuWorkerHostNames
+			}
+			*patches = append(*patches, hostNamesPatch)
+		}
+	}
+}
+
+// add TPU_WORKER_HOSTNAMES to containers in a ray cluster
+func mutateRayCluster(admissionReview *admissionv1.AdmissionReview) (*admissionv1.AdmissionResponse, error) {
+	raycluster, err := extractRayCluster(admissionReview)
+	var patches []patch
+
+	if err == nil {
```

Comment:
I would invert this and check the error immediately,  to avoid putting all your main logic inside an indented section, and having an unnecessary "else."

```go
raycluster, err := extractRayCluster(admissionReview)
if err != nil {
    klog.Fatalf("Ray Cluster extraction failed: %s", err)
}

var patches []patch
for i := 0; i < len(raycluster.Spec.WorkerGroupSpecs); i++ {
   ...
}
```